Add verified properties to event definitions

alexkim205 opened the pull request:
## Changes

Context https://github.com/PostHog/posthog/issues/7991. "Verifying" an event will be a manual action that a user can take via the app that helps teams taxonomize/prioritize certain events over others.

Simple PR that adds a new "verified" property to event definitions. 

I don't believe this requires plugin-server tests/changes since verification cannot happen during ingestion, but @pauldambra please lmk I'm overlooking something.

## How did you test this code?

Running api locally and adding tests. Also checked on Data grip that column is backfilled properly after migrating.

<img width="457" alt="Screen Shot 2022-01-14 at 12 40 22 PM" src="https://user-images.githubusercontent.com/13460330/149582777-b212c7f2-dee8-4726-85b0-cc43f6e8c369.png">

alexkim205 committed:
add verified boolean property to event definitions

alexkim205 committed:
fix tests

alexkim205 committed:
don't make nullable

paolodamico commented on posthog/models/event_definition.py:
```diff
@@ -19,6 +19,7 @@ class EventDefinition(UUIDModel):
     query_usage_30_day: models.IntegerField = models.IntegerField(
         default=None, null=True,
     )  # Number of times the event has been used in a query in the last 30 rolling days (computed asynchronously)
+    verified: models.BooleanField = models.BooleanField(default=False, null=False)
```
I wonder if we should consider something like `verified_at` & `verified_by` instead of just a `boolean`, seems like very valuable information to show on the UI.

alexkim205 replied:
that's a good idea. Agree w/ verified_by but I think verified_at would become repetitive metadata if we implement some kind of general insight history

alexkim205 committed:
add blank true so that verified doesnt need to be passed down

paolodamico reviewed (dismissed):
Made a suggestion on how to track this, but your call, standalone, this PR lgtm.

alexkim205 committed:
add verified, verified_at, verified_by

alexkim205 committed:
fix conditional

alexkim205 committed:
fix tests

pauldambra reviewed (dismissed):
Better to the name the migration than keep the auto generated name

Other than that LGTM

paolodamico commented on ee/api/ee_event_definition.py:
```diff
@@ -21,6 +22,9 @@ class Meta:
             "updated_at",
             "updated_by",
             "last_seen_at",
+            "verified",
```
Seems like `verified` is redundant if we have `verified_at`, but unsure if we'd plan to eventually get rid of that attribute with the version control history.

alexkim205 replied:
Yup that's what I had planned as well. `verified_at` will be eventually deprecated once version control becomes a thing.

paolodamico commented on ee/api/ee_event_definition.py:
```diff
@@ -30,10 +34,15 @@ class Meta:
     def update(self, event_definition: EnterpriseEventDefinition, validated_data):
         validated_data["updated_by"] = self.context["request"].user
+        if "verified" in validated_data and validated_data["verified"]:
+            validated_data["verified_by"] = self.context["request"].user
```
- This could lead to the verified_by being overridden even if the event has already been verified.
- If the event was previously verified and you remove verification we should clear `verified_at` & `verified_by`

Having said the above, if we do plan to have full history on this soon as you suggested, we then might be better off removing all these.

alexkim205 replied:
I'll move clearing logic in the model up into the serializer, which should take care of both points. I'd prefer to separate this feature from full history, but we can consider removing this once history is fully implemented

paolodamico commented on ee/models/event_definition.py:
```diff
@@ -10,3 +12,26 @@ class EnterpriseEventDefinition(EventDefinition):
+        super().__init__(*args, **kwargs)
+        if self.verified:
+            self.verified_at = timezone.now()
+        self._prev_verified = self.verified
```
Not sure if this is the best way to check if the event was previously verified. See suggestion below.

paolodamico commented on ee/models/event_definition.py:
```diff
@@ -10,3 +12,26 @@ class EnterpriseEventDefinition(EventDefinition):
+
+    def save(self, *args, **kwargs):
+        # Previously unverified event becomes verified
+        if not self._prev_verified and self.verified:
```
```suggestion
        if not EnterpriseEventDefinition.objects.get(pk=self.pk).verified and self.verified:
```

alexkim205 replied:
I'm curious, what is the advantage of using the ORM over caching the previous value? It seems cache would be slightly faster here.

alexkim205 replied:
Regardless, it makes sense to do it your way if this logic is bubbled up into the serializer.

paolodamico replied:
Well I'm not sure the cache would be faster, because the ORM is lazy not sure exactly when will it fetch that value, but I think regardless it'll be the same total number of queries for the endpoint.

paolodamico commented on ee/models/event_definition.py:
```diff
@@ -10,3 +12,26 @@ class EnterpriseEventDefinition(EventDefinition):
+            self.verified_at = timezone.now()
+        # Event becomes unverified
+        elif not self.verified:
+            self.verified_at = None
```
Perhaps we should have this logic in the serializer so everything is in one place?

alexkim205 replied:
good idea

paolodamico commented on ee/api/test/test_event_definition.py:
```diff
@@ -113,7 +116,79 @@ def test_with_expired_license(self):
+                team=self.team, name="enterprise event", owner=self.user, verified=None
+            )
+
+    def test_errors_on_invalid_verified_by_type(self):
```
This is not an API test, should go some place else.

paolodamico reviewed (commented):
Some suggestions inline.

pauldambra commented:
> I don't believe this requires plugin-server tests/changes since verification cannot happen during ingestion

I guess any test which asserts on event definitions might start failing because you'd know have `verified: false` ??

But it seems to be green now :)

alexkim205 commented:
> > I don't believe this requires plugin-server tests/changes since verification cannot happen during ingestion
> 
> I guess any test which asserts on event definitions might start failing because you'd now have `verified: false` ??
> 
> But it seems to be green now :)

Did some digging bc I suspected event def tests to fail too but it looks like it was tripping up on other random tests. Not sure what happened there but I'll take it 🤷

alexkim205 committed:
add more tests, refactor logic, rename migration

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-event-verified-property

alexkim205 commented:
@pauldambra @paolodamico ready for a rereview. In this revision I added a few more tests, abstracted all logic into the serializer, and renamed the migration.

alexkim205 committed:
move model test to ee

pauldambra commented on ee/api/ee_event_definition.py:
```diff
@@ -30,13 +35,39 @@ class Meta:
+            validated_data["verified_at"] = None
+            validated_data["verified"] = False
+        else:
+            # Don't allow editing verified properties in any other situation
```
May be my understanding of Django. Would this allow a signed in user to maliciously or accidentally unverify events just by not sending "verified" in the API call?

pauldambra replied:
Should we just ignore verified fields in this circumstance?

paolodamico replied:
Yes, I've done some cleanup, pushing changes in a sec, finishing up the review.

alexkim205 replied:
Thanks for taking care of this :)

pauldambra reviewed (approved):
Happy for API comment to be sorted in a follow-up (if I'm not confused about how it could be taken advantage of)

paolodamico committed:
minor cleanup

paolodamico reviewed (approved):
did some minor cleanup on the serializer to make it clearer and address an edge case, let's merge this 💪

paolodamico committed:
Update test_event_definition.py
